### PR TITLE
feat: New functions and operations for working with arrays

### LINKED ISSUE #6075: Add array expressions
### Is your feature request related to a problem or challenge?

Hello,

I have an array data type and i would like to be able to run the following functions.

### Describe the solution you'd like

array_length(col)
array_contains(col<T>, T) `AND/OR` select * from my_table where 'text'=ANY(my_col);
array_unnest()

```
# TABLE
| food | ingredients          |
| ---- | -------------------- |
| user | [eggs, salt, pepper] |

# SQL INPUT
# SELECT food, unnest(ingredients) from TABLE

# SQL OUTPUT
| food | unnest(ingredients) |
| ---- | ------------------- |
| user | eggs                |
| user | pepper              |
| user | salt                |
```

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### LINKED ISSUE #6119: New functions and operations for working with arrays
### Is your feature request related to a problem or challenge?

Arrow-DataFusion can support creating an array of a certain data type (`make_array`), but there are no functions to work with them.
In my opinion, the documentation of the database PostgreSQL can serve as a reference for working on arrays.

The following features are considered for implementation:
Operators:
- [ ] Any

Array operators:
- [ ] concatenation
- [ ] overlap
- [ ] concatenate

Array functions:
- [ ] array_append
- [ ] array_cat
- [ ] array_ndims
- [ ] array_dims
- [ ] array_fill
- [ ] array_length
- [ ] array_lower
- [ ] array_position
- [ ] array_positions
- [ ] array_prepend
- [ ] array_remove
- [ ] array_replace
- [ ] array_to_string
- [ ] array_upper
- [ ] cardinality
- [ ] string_to_array
- [ ] unnest

### Describe the solution you'd like

Series of PR, active discussion of features.

### Describe alternatives you've considered

The significance of some features can be disproven as a result of conversations.

### Additional context

Sources:
The array documentation: [PostgreSQL](https://postgrespro.ru/docs/postgrespro/15/functions-array?lang=en)
#6073
#6075

### PR DESCRIPTION
**Pull request by izveigor**
# Which issue does this PR close?

Closes #6119 
Closes #6075.

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?
Yes

# Are there any user-facing changes?
Yes

### CODE REVIEW

**Comment by alamb** [comment]
Thank you @izveigor  -- I have put this on my review list but I likely won't have a chance to review until tomorrow

**Comment by alamb** [comment]
I didn't make it to this today, but I plan to review it tomororw

File: datafusion/core/tests/sqllogictests/test_files/array.slt [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+#############
+## Array expressions Tests
+#############
+
+# array scalar function #1
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/array.slt:22]: These are great @izveigor  -- thank you so much

the only thing I recommend is adding some additional tests that have `null` in the lists.

File: datafusion/expr/src/function.rs [hunk 2]
```diff
@@ -322,10 +404,21 @@ pub fn signature(fun: &BuiltinScalarFunction) -> Signature {
 
     // for now, the list is small, as we do not have many built-in functions.
     match fun {
-        BuiltinScalarFunction::MakeArray => Signature::variadic(
-            array_expressions::SUPPORTED_ARRAY_TYPES.to_vec(),
-            fun.volatility(),
-        ),
+        BuiltinScalarFunction::ArrayAppend => Signature::any(2, fun.volatility()),
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/expr/src/function.rs:407]: 🤔  Given the element type of the list is part of its `DataType`, you probably can'y use the existing Signatures

Perhaps you could add a new `Signature::any_list` or something that would only check that the datatype matched `DataType::list` 🤔

[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/expr/src/function.rs:407]: 🤔  Given the element type of the list is part of its `DataType`, you probably can'y use the existing Signatures

Perhaps you could add a new `Signature::any_list` or something that would only check that the datatype matched `DataType::list` 🤔

File: datafusion/physical-expr/src/array_expressions.rs [hunk 3]
```diff
@@ -77,11 +81,40 @@ fn array_array(args: &[ArrayRef]) -> Result<ArrayRef> {
...
         return Err(DataFusionError::Internal(
-            "array requires at least one argument".to_string(),
+            "Array requires at least one argument".to_string(),
         ));
     }
 
-    let res = match args[0].data_type() {
+    let data_type = args[0].data_type();
+    let res = match data_type {
+        DataType::List(..) => {
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/physical-expr/src/array_expressions.rs:90]: As FixedSizedList and List are different data types, if people have data that came from a Parquet file or something that is a FixedSizedList these functions likely wont work, 

However, perhaps eventually we can add coercion rules to coerce (automatically cast) FixedSizeList to List

File: datafusion/physical-expr/src/array_expressions.rs [hunk 4]
```diff
@@ -77,11 +81,40 @@ fn array_array(args: &[ArrayRef]) -> Result<ArrayRef> {
...
+            let arrays =
```
Comment (alamb) [inline datafusion/physical-expr/src/array_expressions.rs:91]: @tustvold  can you offer some suggestions on using the arrow-rs API to build list arrays? Is this the best way to use that API?

File: datafusion/physical-expr/src/array_expressions.rs [hunk 5]
```diff
@@ -115,3 +149,1560 @@ pub fn array(values: &[ColumnarValue]) -> Result<ColumnarValue> {
...
+/// Array_append SQL function
+pub fn array_append(args: &[ColumnarValue]) -> Result<ColumnarValue> {
+    if args.len() != 2 {
+        return Err(DataFusionError::Internal(format!(
+            "Array_append function requires two arguments, got {}",
+            args.len()
+        )));
+    }
+
+    let arr = match &args[0] {
```
Comment (alamb) [inline datafusion/physical-expr/src/array_expressions.rs:189]: I think you can use https://docs.rs/datafusion/latest/datafusion/physical_plan/enum.ColumnarValue.html#method.into_array here

File: datafusion/physical-expr/src/array_expressions.rs [hunk 6]
```diff
@@ -115,3 +149,1560 @@ pub fn array(values: &[ColumnarValue]) -> Result<ColumnarValue> {
...
+            scalars.push(ColumnarValue::Scalar(ScalarValue::try_from_array(
+                &concat, i,
+            )?));
+        }
+        scalars
+    }};
+}
+
+/// Array_append SQL function
+pub fn array_append(args: &[ColumnarValue]) -> Result<ColumnarValue> {
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/physical-expr/src/array_expressions.rs:181]: The difference is that if you take `ColumnarValue` we could specialize the kernels to do something faster with scalar (single) values rather than expanding them out to arrays (aka making copies).

For the initial implementation I think converting them all to arrays is the best approach as it is simplest

File: datafusion/physical-expr/src/functions.rs [hunk 7]
```diff
@@ -2785,73 +2807,6 @@ mod tests {
         Ok(())
     }
 
-    fn generic_test_array(
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/physical-expr/src/functions.rs:2788]: I am not sure

**Review by alamb — COMMENTED** [review]
This PR looks really nice @izveigor  -- thank you so much!

I haven't had a chance to review all the function implementations yet but the overall structure looks great to me . I am hoping to get @tustvold  or someone else who is more of an expert in the arrow-rs structures here to offer an opinion on the structure of the kernels

I'll try and complete my review soon

File: datafusion/physical-expr/src/array_expressions.rs [hunk 8]
```diff
@@ -115,3 +149,1560 @@ pub fn array(values: &[ColumnarValue]) -> Result<ColumnarValue> {
...
+        .iter()
+        .map(|x| match x {
+            ColumnarValue::Array(array) => array.clone(),
+            ColumnarValue::Scalar(scalar) => scalar.to_array().clone(),
+        })
+        .collect();
+    let data_type = arrays[0].data_type();
+    match data_type {
+        DataType::List(..) => {
+            let list_arrays =
```
Comment (tustvold) [inline datafusion/physical-expr/src/array_expressions.rs:322]: I think you could just call `to_data()` I'm not sure this needs to downcast to ListArray

File: datafusion/physical-expr/src/array_expressions.rs [hunk 9]
```diff
@@ -115,3 +149,1560 @@ pub fn array(values: &[ColumnarValue]) -> Result<ColumnarValue> {
...
+                downcast_vec!(arrays, ListArray).collect::<Result<Vec<&ListArray>>>()?;
+            let len: usize = list_arrays.iter().map(|a| a.values().len()).sum();
+            let capacity = Capacities::Array(
+                list_arrays.iter().map(|a| a.get_buffer_memory_size()).sum(),
```
Comment (tustvold) [inline datafusion/physical-expr/src/array_expressions.rs:326]:
```suggestion
                list_arrays.iter().map(|a| a.len()).sum(),
```

File: datafusion/physical-expr/src/array_expressions.rs [hunk 10]
```diff
@@ -115,3 +149,1560 @@ pub fn array(values: &[ColumnarValue]) -> Result<ColumnarValue> {
...
+                "Array is not type '{data_type:?}'."
+            )))
+        }
+    };
+
+    array(arrays.as_slice())
+}
+
+/// Array_concat/Array_cat SQL function
+pub fn array_concat(args: &[ColumnarValue]) -> Result<ColumnarValue> {
```
Comment (tustvold) [inline datafusion/physical-expr/src/array_expressions.rs:311]: It would perhaps be nicer to use a combination of https://docs.rs/arrow-array/latest/arrow_array/array/struct.GenericListArray.html#method.try_new and https://docs.rs/arrow-select/latest/arrow_select/concat/index.html

File: datafusion/physical-expr/src/array_expressions.rs (hunk 4 above)
Comment (tustvold) [reply datafusion/physical-expr/src/array_expressions.rs:91]: It would perhaps be nicer to use a combination of https://docs.rs/arrow-array/latest/arrow_array/array/struct.GenericListArray.html#method.try_new and https://docs.rs/arrow-select/latest/arrow_select/concat/index.html

MutableArrayData is not the nicest API to use

**Comment by alamb** [comment]
> What do you think, @alamb?

I think this would be ok -- especially as you have a history of continued contribution. However, there are a few instances where engnaged committers committed in the start of promising features (such as the analysis framework from @isidentical) and then were not able to to finish the work for whatever reason. While this is fine, I think it would be better for datafusion to avoid it.

Thus I would like to suggest an alternate approach which is to break this PR down into several smaller ones (perhaps one for each new function?) That way we can give each function the attention during review it deserves (and maybe even parallelize the work)

We have a much better track record of being able to review and merge smaller PRs quickly than single large PRs. So when the functionality can be split up I think that is the best plan.

What do you think @izveigor ?

**Comment by alamb** [comment]
> This PR is completed, and the next PRs will be only improve it.

I agree this PR is complete (with tests) and is not missing anything major

> If we break this PR down, I think it will increase production time.

Yes, I agree breaking the PR down will require more effort on the author's (your) part. However, I do think if you have the time the effort would improve the overall quality of the DataFusion codebase. 

> What do you think, @alamb?

I think we can merge this PR as long as the work you have planned is tracked by some tickets (so that if you don't have a chance to get to them at least we will have some institutional knowledge) 

Is that acceptable?

**Comment by alamb** [comment]
> I have created issues regarding further improvements for working with arrays:


Thanks @izveigor  -- I added them to https://github.com/apache/arrow-datafusion/issues/2326 as well. !

**Comment by jackwener** [comment]
This PR exist bug, related with #6596
